htlcswitch/switch.go

```diff
@@ -300,11 +309,39 @@ func (s *Switch) handleLocalDispatch(payment *pendingPayment, packet *htlcPacket
 	case *lnwire.UpdateFailHTLC:
 		// Retrieving the fail code from byte representation of error.
 		var userErr error
-		if code, err := htlc.Reason.ToFailCode(); err != nil {
-			userErr = errors.Errorf("can't decode fail code id"+
-				"(%v): %v", htlc.ID, err)
+
+		failure, err := payment.deobfuscator.Deobfuscate(htlc.Reason)
+		if err != nil {
+			userErr = errors.Errorf("unable to de-obfuscate "+
+				"onion failure, htlc with id(%v): %v", htlc.ID, err)
+			log.Error(userErr)
 		} else {
-			userErr = errors.New(code)
+			// Process payment failure by updating the lightning network
+			// topology by using router subsystem handler.
+			var update *lnwire.ChannelUpdate
+
+			switch failure := failure.(type) {
+			case *lnwire.FailTemporaryChannelFailure:
+				update = failure.Update
+			case *lnwire.FailAmountBelowMinimum:
+				update = &failure.Update
+			case *lnwire.FailFeeInsufficient:
+				update = &failure.Update
+			case *lnwire.FailIncorrectCltvExpiry:
+				update = &failure.Update
+			case *lnwire.FailExpiryTooSoon:
+				update = &failure.Update
+			case *lnwire.FailChannelDisabled:
+				update = &failure.Update
+			}
+
+			if update != nil {
+				log.Info("Received payment failure(%v), applying lightning "+
+					"network topology update", failure.Code())
+				s.cfg.UpdateTopology(update)
```

Comment:
We could completely eliminate the requirement of the switch to _manually_ update the directed edge info when it encounters a payment error. Rather than up calling into the router to update the information, it could instead just return the error as is. However, it wouldn't be a typical error, instead the router would have a special `PaymentError` type. When a payment errors out, it can perform a type switch on the error. If it's this particular error, then it can decide exactly what to do with it based on the returned `lnwire.FailureCode`. Optionally an update would also be attached. 

The above method allows the router complete control over its actions when it receive an error. For example, if we send a payment and receive an error that says our fee is out of date, then can modify the route, and immediately re-send the payment. Other situations would be handled case-by-case. 

Note that this isn't a blocker for the PR, if you opt not to implement it, it can be done in some follow up work by either yourself or me.